src/Microsoft.Data.Analysis/DataFrame.IO.cs

```diff
@@ -127,70 +197,63 @@ private static Type MaxKind(Type a, Type b)
 
             var linesForGuessType = new List<string[]>();
             long rowline = 0;
-            int numberOfColumns = 0;
+            int numberOfColumns = dataTypes != null ? dataTypes.Length : 0;
 
             if (header == true && numberOfRowsToRead != -1)
                 numberOfRowsToRead++;
 
+            DataFrame ret;
             List<DataFrameColumn> columns;
             long streamStart = csvStream.Position;
             // First pass: schema and number of rows.
             using (var streamReader = new StreamReader(csvStream, encoding: null, detectEncodingFromByteOrderMarks: true, bufferSize: -1, leaveOpen: true))
             {
-                string line = streamReader.ReadLine();
-                while (line != null)
+                string line = null;
+                if (dataTypes == null)
                 {
-                    if ((numberOfRowsToRead == -1) || rowline < numberOfRowsToRead)
+                    line = streamReader.ReadLine();
+                    while (line != null)
                     {
-                        if (linesForGuessType.Count < guessRows)
+                        if ((numberOfRowsToRead == -1) || rowline < numberOfRowsToRead)
                         {
-                            var spl = line.Split(separator);
-                            if (header && rowline == 0)
+                            if (linesForGuessType.Count < guessRows)
                             {
-                                if (columnNames == null)
-                                    columnNames = spl;
-                            }
-                            else
-                            {
-                                linesForGuessType.Add(spl);
-                                numberOfColumns = Math.Max(numberOfColumns, spl.Length);
+                                var spl = line.Split(separator);
+                                if (header && rowline == 0)
+                                {
+                                    if (columnNames == null)
+                                        columnNames = spl;
+                                }
+                                else
+                                {
+                                    linesForGuessType.Add(spl);
+                                    numberOfColumns = Math.Max(numberOfColumns, spl.Length);
+                                }
                             }
                         }
+                        ++rowline;
+                        if (rowline == guessRows)
```

Comment:
This line ensures that we read in up to `guessRows` number of rows to perform our type inferencing, no more. Previously, this could have been the entire file.